File: Tools/Source/tempConv.py
# TEMPERATURE CONVERSION CHART -- V.PY
# William Wadsworth
# CSC1710
# Created: 1.10.2020
# 
# Python-ized: 3.11.2024
# Updated 8.17.2024: PEP 8 Compliance
# 
# This program creates a temperature conversion chart based on a degree given
# in Fahrenheit, incrementing by a value imput by the user.
# 
# Usage: python3 tempConv.py


from sys import exit


def main():
    # --- INTRODUCTION --------------------------
    print("""This program creates a temperature convrsion chart based on a degree given in
    Fahrenheit, incrementing by a value you choose.\nAll values must be rounded to the nearest 
    thousandth.""")

    # --- CONFIRMATION --------------------------
    confirmation = input("Do you want to run this program? [Y/n]: ")
    # check confirmation
    while confirmation not in ("Y","n"):
        confirmation = input("Please enter [Y/n]: ")
    # if declined, terminate
    if confirmation == "n":
        print('terminating...')
        exit(0)

    # --- SMALLEST DEGREE -----------------------
    start_degree = int(input("Give your starting (smallest) Fahrenheit degree [-1000 < this_degree < 1000]: "))
    # input validation
    while start_degree < -1000 or start_degree > 1000:
        start_degree = int(input("Not valid, degree limitations: [-1000 < this_degree < 1000]: "))

    # --- LARGEST DEGREE ------------------------
    end_degree = int(input("Give your ending (largest) Fahrenheit degree [smallest_degree < this_degree < 1000]: "))
    # input validation
    while end_degree <= start_degree or end_degree > 1000:
        start_degree = int(input("Not valid, degree limitations: [smallest_degree < this_degree < 1000]: "))

    # --- INCREMENT -----------------------------
    increment = float(input("How much do you want to increment by: "))
    # input validation
    while increment <= 0:
        increment = float(input("Not valid, increment must be > 0: "))

    # --- TABLE + FORMULAS ----------------------
    print(""" Fahrenheit (°F) |  Celsius (°C)  |   Kelvin (K)   
    ---------------------------------------------------""")

    # celsius and kelvin formulae:
    # float cel = ((start_degree -32) * 5/9)
    # float kel = ((start_degree -32) * 5/9 + 273.15)

    # while loop to run through incremented degrees
    current_degree = float(start_degree)
    while start_degree <= end_degree:
        # display calculations
        celsius: float = (current_degree - 32) * 5 / 9
        kelvin: float = celsius + 273.15
        print(f"{current_degree:12.2f}     |{celsius:12.2f}    |{kelvin:12.2f}")
        current_degree += increment


if __name__ == "__main__":
    main()

File: Tools/Source/tempConv.rs
/* TEMPERATURE CONVERSION CHART -- V.Rust
 * William Wadsworth
 * CSC1710
 * Created: 1.10.2020
 * Rust-ified: 3.20.2024
 * 
 * This program creates a temperature conversion chart based on a degree given in Fahrenheit,
 * incrementing by a value imput by the user.
 * 
 * Usage: 
 * To compile: rustc tempConv.rs
 * To run: ./tempConv
*/

use std::io::{self, Write};
use std::process;

//fn main() -> io::Result<()> 
fn main() {
    // introduction
    println!("This program creates a temperature conversion chart based on a degree given in \
             Fahrenheit, incrementing by a value you choose.\nAll values must be rounded to the \
             nearest thousandth.\n");

    // confirmation
    let mut confirmation = String::new();

    print!("Do you want to run this program? [Y/n]: ");
    io::stdout().flush().expect("Failed to flush stdout");
    loop { // sort of not really like cpp do-while
        confirmation.clear(); // clear input before reading again
        io::stdin().read_line(&mut confirmation).expect("Failed to read response");

        match confirmation.trim().to_lowercase().as_str() {
            "y" | "yes" => {
                break;
            }
            "n" | "no" => {
                println!("Exiting...");
                process::exit(0);
            }
            _ => { // invalid value
                print!("Invalid input. Please enter 'Y' or 'n': ");
                io::stdout().flush().expect("Failed to flush stdout");
            }
        }
    }

    // --- SMALLLEST DEGREE ---------------------
    let mut input = String::new();
    let smallest_degree: f64;

    // give instructions to define sdegree
    print!("Give your starting (smallest) Fahrenheit degree between -1000 and 1000: ");
    io::stdout().flush().expect("Failed to flush stdout");
    loop {
        input.clear();
        io::stdin().read_line(&mut input).expect("Failed to read degree");

        // input validation
        match input.trim().parse::<f64>() {
            Ok(value) if (-1000.0..=1000.0).contains(&value) => {
                smallest_degree = value;
                break;
            }
            _ => {
                print!("Not valid, degree limitations: [-1000 < this_degree < 1000]: ");
                io::stdout().flush().expect("Failed to flush stdout");
            }
        }
    }

    // --- LARGEST DEGREE -----------------------
    // same thing as smallest_degree
    let largest_degree: f64;

    print!("{}", format!("Give your ending (largest) Fahrenheit degree between {} and 1000: ", smallest_degree));
    io::stdout().flush().expect("Failed to flush stdout");
    loop {
        input.clear();
        io::stdin().read_line(&mut input).expect("Failed to read degree");

        match input.trim().parse::<f64>() {
            Ok(value) if (smallest_degree..=1000.0).contains(&value) => {
                largest_degree = value;
                break;
            }
            _ => {
                print!("{}", format!("Not valid, degree limitations: [{} < this_degree < 1000]: ", smallest_degree));
                io::stdout().flush().expect("Failed to flush stdout");
            }
        }
    }

    // --- INCREMENT ----------------------------
    // guess what? same thing again
    let increment: f64;

    print!("How much do you want to increment by: ");
    io::stdout().flush().expect("Failed to flush stdout");
    loop {
        input.clear();
        io::stdin().read_line(&mut input).expect("Failed to read increment");

        match input.trim().parse::<f64>() {
            Ok(value) if value > 0.0 && value < (largest_degree-smallest_degree) => {
                increment = value;
                break;
            }
            Ok(_) => { // successful parsing, but invalid value
                print!("{}", format!("Invalid, increment must be (0 < increment < {}): ", largest_degree-smallest_degree));
                io::stdout().flush().expect("Failed to flush stdout");
            }
            Err(_) => { // failure case
                print!("{}", format!("Invalid, increment must be (0 < increment < {}): ", largest_degree-smallest_degree));
                io::stdout().flush().expect("Failed to flush stdout");
            }
        }
    }

    // --- TABLE AND FORMULAS -------------------
    // make table heading using spaces
    println!(" Fahrenheit (°F) |  Celsius (°C)  |   Kelvin (K)   ");
    println!("---------------------------------------------------");

    // define celsius and kelvin formulae
    //double c = ((sdegree - 32) * 5/9);
    //double k = ((sdegree - 32) * 5/9 + 273.15);

    // while loop to run through incremented degrees
    let mut current_value: f64 = smallest_degree;

    while current_value <= largest_degree {
        // display calculations
        println!(
            "{:>12.3}     |{:>12.3}    |{:>12.3}",
            current_value,
            (current_value -32.0) * 5.0 / 9.0,
            (current_value -32.0) * 5.0 / 9.0 + 273.15
        );
        current_value += increment;
    }
}

